### PR TITLE
Add support fuzzy matching a list of dictionaries using either a key or a callback function

### PR DESCRIPTION
**Pull request by yegappan**

### CODE REVIEW

File: runtime/doc/eval.txt [hunk 1]
```diff
@@ -7309,12 +7311,24 @@ matchend({expr}, {pat} [, {start} [, {count}]])			*matchend()*
 			GetText()->matchend('word')
 
 
-matchfuzzy({list}, {str})			*matchfuzzy()*
-		Returns a list with all the strings in {list} that fuzzy
-		match {str}. The strings in the returned list are sorted
-		based on the matching score. {str} is treated as a literal
-		string and regular expression matching is NOT supported.
-		The maximum supported {str} length is 256.
+matchfuzzy({list}, {str} [, {key}])
```
Comment (prabirshrestha) [inline runtime/doc/eval.txt]: based on my comment here might be good to not support key but instead support expression with `v:val` which can be optimized to extract key. https://github.com/vim/vim/pull/6932#issuecomment-691677822

**Review by prabirshrestha — APPROVED** [review]

File: src/testdir/test_functions.vim [hunk 2]
```diff
@@ -2576,6 +2578,30 @@ func Test_matchfuzzy()
...
+  call assert_fails("let x = matchfuzzy(l, 'foo', {'key' : []})", 'E730:')
+  call assert_fails("let x = matchfuzzy(l, 'cam', test_null_dict())", 'E715:')
+  call assert_fails("let x = matchfuzzy(l, 'foo', {'key' : test_null_string()})", 'E475:')
+  call assert_fails("let x = matchfuzzy(l, 'foo', {'text_cb' : test_null_function()})", 'E475:')
+
+  let l = [{'id' : 5, 'name' : 'foo'}, {'id' : 6, 'name' : []}, {'id' : 7}]
+  call assert_fails("let x = matchfuzzy(l, 'foo', {'key' : 'name'})", 'E730:')
+
+  " Test for the fuzzymatchpos() function
+  call assert_equal([['curl', 'world'], [[2,3], [2,3]]], matchfuzzypos(['world', 'curl'], 'rl'))
```
Comment (prabirshrestha) [inline src/testdir/test_functions.vim]: missing tests for key and text_cb for matchfuzzypost

Comment (prabirshrestha) [inline src/testdir/test_functions.vim]: can we also have a test case for multiple matches in the same item.

```vim
matchfuzzypos(['hello world hello world', 'hello', 'world]`, 'hello')
```

File: src/testdir/test_functions.vim [hunk 3]
```diff
@@ -2569,13 +2571,44 @@ func Test_matchfuzzy()
...
+  call assert_fails("let x = matchfuzzy(l, 'cam', test_null_dict())", 'E715:')
+  call assert_fails("let x = matchfuzzy(l, 'foo', {'key' : test_null_string()})", 'E475:')
+  call assert_fails("let x = matchfuzzy(l, 'foo', {'text_cb' : test_null_function()})", 'E475:')
+
+  let l = [{'id' : 5, 'name' : 'foo'}, {'id' : 6, 'name' : []}, {'id' : 7}]
+  call assert_fails("let x = matchfuzzy(l, 'foo', {'key' : 'name'})", 'E730:')
+endfunc
+
+" Test for the fuzzymatchpos() function
+func Test_matchfuzzypos()
```
Comment (prabirshrestha) [inline src/testdir/test_functions.vim]: missing tests for empty array. `matchfuzzypos([], 'foo')`

File: runtime/doc/eval.txt [hunk 4]
```diff
@@ -7325,11 +7341,26 @@ matchfuzzy({list}, {str})			*matchfuzzy()*
...
+		   :echo getbufinfo()->matchfuzzy("spl",
+						\ {'text_cb' : {v -> v.name}})
+<		results in a list of buffer information dicts with buffer
+		names fuzzy matching "spl". >
 		   :echo v:oldfiles->matchfuzzy("test")
 <		results in a list of file names fuzzy matching "test". >
 		   :let l = readfile("buffer.c")->matchfuzzy("str")
 <		results in a list of lines in "buffer.c" fuzzy matching "str".
 
+matchfuzzypos({list}, {str} [, {dict}])			*matchfuzzypos()*
```
Comment (prabirshrestha) [inline runtime/doc/eval.txt:7358]: add example for dict or might be add a comment to refer to matchfuzzy dict option.

File: src/testdir/test_functions.vim [hunk 5]
```diff
@@ -2569,13 +2571,44 @@ func Test_matchfuzzy()
...
+  call assert_equal([['curl', 'world'], [[2,3], [2,3]]], matchfuzzypos(['world', 'curl'], 'rl'))
+  call assert_equal([['curl', 'world'], [[2,3], [2,3]]], matchfuzzypos(['world', 'one', 'curl'], 'rl'))
```
Comment (prabirshrestha) [inline src/testdir/test_functions.vim]: shouldn't this response be this instead? so that if there are multiple highlights it can still be represented correctly?

```json
[
   [ "curl", "world", "curl world"],
   [
       [ [2,3] ],
       [ [2,3 ] ],
       [  [2,3], [7,8] ]
   ]
]
```

**Review by prabirshrestha — CHANGES_REQUESTED** [review]
added few comments.

**Comment by vim-ml** [comment]
Hi,

On Tue, Sep 15, 2020 at 10:50 AM Prabir Shrestha <vim-dev-github@256bit.org>
wrote:

> *@prabirshrestha* requested changes on this pull request.
>
> added few comments.
> ------------------------------
>
> In src/testdir/test_functions.vim
> <https://github.com/vim/vim/pull/6947#discussion_r488855126>:
>
> > +  call assert_fails("let x = matchfuzzy(l, 'day', {'text_cb' : {a, b -> 1}})", 'E119:')
> +  call assert_equal([], matchfuzzy(l, 'cam'))
> +  call assert_fails("let x = matchfuzzy(l, 'cam', {'text_cb' : []})", 'E921:')
> +  call assert_fails("let x = matchfuzzy(l, 'foo', {'key' : []})", 'E730:')
> +  call assert_fails("let x = matchfuzzy(l, 'cam', test_null_dict())", 'E715:')
> +  call assert_fails("let x = matchfuzzy(l, 'foo', {'key' : test_null_string()})", 'E475:')
> +  call assert_fails("let x = matchfuzzy(l, 'foo', {'text_cb' : test_null_function()})", 'E475:')
> +
> +  let l = [{'id' : 5, 'name' : 'foo'}, {'id' : 6, 'name' : []}, {'id' : 7}]
> +  call assert_fails("let x = matchfuzzy(l, 'foo', {'key' : 'name'})", 'E730:')
> +endfunc
> +
> +" Test for the fuzzymatchpos() function
> +func Test_matchfuzzypos()
> +  call assert_equal([['curl', 'world'], [[2,3], [2,3]]], matchfuzzypos(['world', 'curl'], 'rl'))
> +  call assert_equal([['curl', 'world'], [[2,3], [2,3]]], matchfuzzypos(['world', 'one', 'curl'], 'rl'))
>
> shouldn't this response be this instead? so that if there is multiple
> highlights it can still be represented correctly?
>
> [
>    [ "curl", "world", "curl world"],
>    [
>        [ [2,3] ],
>        [ [2,3 ] ],
>        [  [2,3], [7,8] ]
>    ]
> ]
>
>
> The fuzzy matching currently looks for only the first match. If there are
multiple matches in the same
string, it currently returns only the first match.

- Yegappan

**Comment by prabirshrestha** [comment]
@yegappan any plans to return multiple highlights for an item? Is it currently the limitation in the library?
Would be good if it supported it even if the library didnt support it so in future can update the library without updating the public api.

**Comment by vim-ml** [comment]
Hi,

On Thu, Sep 17, 2020 at 5:36 PM Prabir Shrestha <vim-dev-github@256bit.org>
wrote:

> @yegappan <https://github.com/yegappan> any plans to return multiple
> highlights for an item? Is it currently the limitation in the library?
> Would be good if it supported it even if the library didnt support it so
> in future can update the library without updating the public api.
>

I am not sure about the use case for returning multiple matches within a
given
string. Why is it not good enough to just highlight the best match within a
string?
How will highlighting multiple matches within a given string help a user?

For example, consider the case of searching for 'aa' in the string
'aaaaaaaa' (worst
case scenario). If all the matches are highlighted, then the entire string
will be
highlighted, which doesn't add much value.

Regards,
Yegappan

**Comment by prabirshrestha** [comment]
It is very common if you are trying to create a fuzzy matching for files.

Here is a real example where I'm searching for `sys\dri\hosts`.  In this case I'm actually interested in `hosts` file and not `sys`. this means `hosts` needs to be highlighted.

image


In projects I can see this using as `controller/feature/featuresA/component.tsx`, `controller/features/featuresA/service.tsx` I might want to use `featureA/service`. which means I want both to highlight.

**Comment by vim-ml** [comment]
Hi,

On Fri, Sep 18, 2020 at 1:23 PM Prabir Shrestha <vim-dev-github@256bit.org>
wrote:

> It is very common if you are trying to create a fuzzy matching for files.
>
> Here is a real example where I'm searching for sys\dri\hosts. In this
> case I'm actually interested in hosts file and not sys. this means hosts
> needs to be highlighted.
>

This is already supported. For example, using your sample input:

let f = ['System32\drivers\etc\lmhosts.sam',
'System32\drivers\mausbhost.sys', 'System32\drivers\etc\hosts.ics',
'System32\drivers\etc\hosts']
echo matchfuzzypos(f, 'Sys\dri\hosts')

produces the following output

[['System32\drivers\etc\hosts', 'System32\drivers\etc\hosts.ics',
'System32\driv
ers\etc\lmhosts.sam', 'System32\drivers\mausbhost.sys'], [[0, 1, 2, 8, 9,
10, 11
, 16, 21, 22, 23, 24, 25], [0, 1, 2, 8, 9, 10, 11, 16, 21, 22, 23, 24, 25],
[0,
1, 2, 8, 9, 10, 11, 16, 23, 24, 25, 26, 27], [0, 1, 2, 8, 9, 10, 11, 16,
22, 23,
 24, 25, 27]]]

Using the matching positions in the above output, you can highlight the
characters like shown
below in your example.

Regards,
Yegappan


> [image: image]
> <https://user-images.githubusercontent.com/287744/93641752-aae6c700-f9b1-11ea-92df-8113c97664b9.png>
>
> In projects I can see this using as
> controller/feature/featuresA/component.tsx,
> controller/features/featuresA/service.tsx I might want to use
> featureA/service. which means I want both to highlight.
>
>
>

**Comment by prabirshrestha** [comment]
If I have 100 matches with each items having 2 highlights and only want to highlight `39:49` items do I need to loop through all the highlights which is 200 iterations or I can just ask for `items[39:49]` and iterate only 20 higlights?

I'm looking something like [ios api](https://developer.apple.com/documentation/uikit/uitableviewdatasource/1614861-tableview?language=objc) for virtualization.

**Comment by vim-ml** [comment]
Hi,

On Fri, Sep 18, 2020 at 3:43 PM Prabir Shrestha <vim-dev-github@256bit.org>
wrote:

> If I want to 100 matches with each items having 2 highlights and only want
> to highlight 39:49 items do I need to loop through all the highlights
> which is 200 iterations or I can just ask for items[39:49] and iterate
> only 20 higlights?
>
>
In the list (second item) returned by matchfuzzypos() you can use list
slices to
get only the matching positions for specific strings. For example, if there
are 100 matching strings and you want to highlight only strings 39 to 49,
then you can slice the second list using 39:49 and use only those items
for highlighting.

- Yegappan


> I'm looking something like ios api
> <https://developer.apple.com/documentation/uikit/uitableviewdatasource/1614861-tableview?language=objc>
> for virtualization.
>
>
>

**Comment by prabirshrestha** [comment]
Now it makes sense. I was looking for range but seems like you are giving character positions so should be possible to do it.

For this text `System32\\drivers\\mausbhost.sys`  when I search for `Sys\dri\hosts` I get  `0, 1, 2, 8, 9, 10, 11, 16, 22, 23,  24, 25, 27`.  If I want to use `prop_add` I need to calculate the length on my own as I most likely want to avoid adding `prop_add` for each character.

```vim
let matches = [ 0, 1, 2, 8, 9, 10, 11, 16, 22, 23, 24, 25, 27 ]

call prop_add(lnum, 0, { 'length': 3 })
call prop_add(lnum, 8, { 'length': 4 })
call prop_add(lnum, 16, { 'length': 1 })
call prop_add(lnum, 22, { 'length': 4 })
call prop_add(lnum, 27, { 'length': 1 })
```

might be what I need now is another function to convert it to range.

```vim
let matches = [ 0, 1, 2, 8, 9, 10, 11, 16, 22, 23, 24, 25, 27 ]
let matchrange = matchpos2range(matches) 
" matchrange = [[0, 3], [8, 4], [16, 1], [22, 4], [27, 1]]

for range in matchrange
    call prop_add(lnum, range[0], { 'length': range[1] })
endfor
```

**Comment by bfrg** [comment]
Here's a very simple command that I used for testing `matchfuzzypos()`:
```vim
if prop_type_get('test')->empty()
    call prop_type_add('test', {'highlight': 'Search'})
endif

function s:fuzzyfind(bang, mod, ...) abort
    if a:0 != 2
        return
    endif

    const files = systemlist('cd ' .. a:1 .. '; find -type f | sed "s/^\.\///"')
    const matches = matchfuzzypos(files, a:2)
    const info = printf('%s [dir: %s]', a:2, fnamemodify(a:1, ':~:.'))

    if empty(matches)
        return
    endif

    silent execute a:mod 'new' fnameescape(info)
    setlocal buftype=nofile noswapfile bufhidden=wipe
    call setline(1, matches[0])

    if a:bang
        for i in len(matches[1])->range()[: winheight(0) - 1]
            call len(matches[1][i])
                    \ ->range()
                    \ ->map({_,j ->
                    \   prop_add(i + 1, matches[1][i][j] + 1, {
                    \     'length': 1,
                    \     'type': 'test',
                    \     'bufnr': bufnr('%')
                    \   })
                    \ })
        endfor
    endif
    setlocal nomodifiable
endfunction

command -nargs=+ -bang -complete=dir Fuzzy call s:fuzzyfind(<bang>0, <q-mods>, <f-args>)
```

Use it like `:Fuzzy! ~/path/to/project/ syshost`.
